File: FSharp.Javascript/Library.js
registerNamespace('FSharp.Javascript.Library.DateTime.TryParse2')


FSharp.Javascript.Library = {
    DateTime: {
        TryParse2: {
            Static: function (x) {
                try {
                    if (x == "") {
                        return new Microsoft.FSharp.Core.FSharpOption.None();
                    }
                    var d = new Date(x)
                    if (d == null) {
                        return new Microsoft.FSharp.Core.FSharpOption.None();
                    }

                    var dateTime = new System.DateTime(d.getFullYear(), d.getMonth() + 1, d.getDate(), d.getHours(), d.getMinutes(), d.getSeconds())
                    return new Microsoft.FSharp.Core.FSharpOption.Some(dateTime);
                }
                catch (err) {
                    return new Microsoft.FSharp.Core.FSharpOption.None();
                }
            }
        }
    },

    Int32: {
        TryParse2: {
            Static: function (x) {
                try {
                    var value = parseInt(x)
                    return new Microsoft.FSharp.Core.FSharpOption.Some(value);
                } catch (err) {
                    return new Microsoft.FSharp.Core.FSharpOption.None();
                }
            }
        }
    },

    Int16: {
        TryParse2: {
            Static: function (x) {
                try {
                    var value = parseInt(x)
                    return new Microsoft.FSharp.Core.FSharpOption.Some(value);
                } catch (err) {
                    return new Microsoft.FSharp.Core.FSharpOption.None();
                }
            }
        }
    },

    Int64: {
        TryParse2: {
            Static: function (x) {
                try {
                    var value = parseInt(x)
                    return new Microsoft.FSharp.Core.FSharpOption.Some(value);
                } catch (err) {
                    return new Microsoft.FSharp.Core.FSharpOption.None();
                }
            }
        }
    },

    Decimal: {
        TryParse2: {
            Static: function (x) {
                try {
                    var value = parseFloat(x)
                    return new Microsoft.FSharp.Core.FSharpOption.Some(value);
                } catch (err) {
                    return new Microsoft.FSharp.Core.FSharpOption.None();
                }
            }
        }
    },

    Double: {
        TryParse2: {
            Static: function (x) {
                try {
                    var value = parseFloat(x)
                    return new Microsoft.FSharp.Core.FSharpOption.Some(value);
                } catch (err) {
                    return new Microsoft.FSharp.Core.FSharpOption.None();
                }
            }
        }
    },

    Boolean: {
        TryParse2: {
            Static: function (x) {
                try {
                    var value = x == "true"
                    var value2 = x == "false"
                    if (value) {
                        return new Microsoft.FSharp.Core.FSharpOption.Some(true);
                    } 
                    else if (value2){
                        return new Microsoft.FSharp.Core.FSharpOption.Some(false);
                    }
                    else {
                        return new Microsoft.FSharp.Core.FSharpOption.None();
                    }
                } catch (err) {
                    return new Microsoft.FSharp.Core.FSharpOption.None();
                }
            }
        }
    }
}

File: FSharp.Javascript/Library.fs
module FSharp.Javascript.Library

open System

type System.DateTime with
    static member TryParse2(x:string) =
        try
            let date = DateTime.Parse(x)
            if date = DateTime.MinValue then 
                None
            else
                Some(date)
        with
        | _ -> None

type System.Int16 with
    static member TryParse2(x:string) =
        try
            let success,result = Int16.TryParse(x)
            if success then Some result else None
        with
        | _ -> None

type System.Int32 with
    static member TryParse2(x:string) =
        try
            let success,result = Int32.TryParse(x)
            if success then Some result else None
        with
        | _ -> None

type System.Int64 with
    static member TryParse2(x:string) =
        try
            let success,result = Int64.TryParse(x)
            if success then Some result else None
        with
        | _ -> None

type System.Decimal with
    static member TryParse2(x:string) =
        try
            let success,result = Decimal.TryParse(x)
            if success then Some result else None
        with
        | _ -> None

type System.Double with
    static member TryParse2(x:string) =
        try
            let success,result = Double.TryParse(x)
            if success then Some result else None
        with
        | _ -> None

type System.Single with
    static member TryParse2(x:string) =
        try
            let success,result = Single.TryParse(x)
            if success then Some result else None
        with
        | _ -> None

type System.Boolean with
    static member TryParse2(x:string) =
        try
            let success,result = Boolean.TryParse(x)
            if success then Some result else None
        with
        | _ -> None

type JObject() =
    let dict = System.Collections.Hashtable()

    member this.Item
        with get(i:string) = dict.[i]
        and set (i:string) (v:obj) = dict.[i] <- v

let (?) (this : JObject) (prop : string) =
    this.[prop]

let (?<-) (this : JObject) (property : string) (value : 'Value) =
    this.[property] <- value